File: 2019/Day06/Solution1.py
import os
from typing import Union, List, Tuple, Dict, Set
from queue import Queue

## **WIP**
class Planet:
    def __init__(self, name: str):
        self.name = name
        self.parent: Union[Planet, None] = None

    def count_orbits(self) -> int:
        if not self.parent:
            return 0
        return 1 + self.parent.count_orbits()

    def path_to_com(self) -> Set[object]:
        distance: Set[Planet] = {}
        queue = Queue()
        queue.put(self)

        while not queue.empty():
            current_node = queue.get()
            distance.add(current_node)

            if current_node.parent:
                queue.put(current_node.parent)

        return distance


def make_system(orbits: List[Tuple[str, str]]) -> Dict[str, Planet]:
    orbital_system = {}
    for parent, child in orbits:
        parent_node: Planet
        child_node: Planet

        if parent in orbital_system.keys():
            parent_node = orbital_system[parent]
        else:
            parent_node = Planet(parent)
            orbital_system[parent] = parent_node

        if child in orbital_system.keys():
            child_node = orbital_system[child]
            child_node.parent = parent_node
        else:
            child_node = Planet(child)
            child_node.parent = parent_node
            orbital_system[child] = child_node

    return orbital_system


def total_orbits(orbital_system: Dict[str, Planet]) -> int:
    totals = [planet.count_orbits() for planet in orbital_system.values()]
    return len(totals)


if __name__ == "__main__":
    directory: str = os.path.dirname(os.path.abspath(__file__))
    filename: str = os.path.join(directory, "input.txt")

    with open(filename, "r", encoding="utf-8") as fp:
        data: List[str] = fp.read().split()

    orbits = [tuple(row.split(")")) for row in data]
    orbital_system = make_system(orbits)
    total = total_orbits(orbital_system)
    print(f"Total orbits is: {total}")

File: 2019/Day06/Solution1.cs
using System;
using System.IO;
using System.Text;
using System.Collections.Generic;
using Interfaces;
using Serilog.Core;
using System.Linq;

namespace Day06
{
    // Planet class that stores the name
    // and parent node that is considered
    // the planet it is orbiting around.
    public class Planet
    {
        public string name;
        public Planet parent;

        public Planet(string name)
        {
            this.name = name;
        }

        // Method traverses through the parents and
        // uses recursion to count the number of parent
        // nodes linked together.
        public int CountOrbits()
        {
            return 1 + parent?.CountOrbits() ?? 0;
        }
    }

    public class Solution1 : ISolution
    {
        // Generates a dictionary of strings associated with Planet objects
        // that have parent and child node relations.
        public Dictionary<string, Planet> GenerateOrbitalMap(
            List<Tuple<string, string>> planets)
        {
            var orbitalSystem = new Dictionary<string, Planet>();
            foreach (var planet in planets)
            {
                Planet parentNode;
                Planet childNode;
                string parent = planet.Item1;
                string child = planet.Item2;

                if (orbitalSystem.ContainsKey(parent))
                {
                    parentNode = orbitalSystem[parent];
                }
                else
                {
                    parentNode = new Planet(parent);
                    orbitalSystem[parent] = parentNode;
                }

                if (orbitalSystem.ContainsKey(child))
                {
                    childNode = orbitalSystem[child];
                    if (childNode.parent == null) childNode.parent = parentNode;
                }
                else
                {
                    childNode = new Planet(child);
                    childNode.parent = parentNode;
                    orbitalSystem[child] = childNode;
                }
            }

            return orbitalSystem;
        }

        // Find the total number of orbits by 
        // traversing through the Planet map.
        public int TotalOrbits(Dictionary<string, Planet> orbits)
        {
            var totals = new List<int>();
            foreach (var planet in orbits.Values)
            {
                totals.Add(planet.CountOrbits());
            }

            return totals.Sum();
        }

        // Takes file input and converts it to a list
        // of Tuples that show parent and child relationships
        // between planet orbits.
        public List<Tuple<string, string>> GenerateInput(string filePath)
        {
            var orbits = new List<Tuple<string, string>>();
            var lines = File.ReadLines(filePath, Encoding.UTF8);
            foreach (string line in lines)
            {
                var planets = line.Split(")");
                orbits.Add(Tuple.Create(planets[0], planets[1]));
            }

            return orbits;
        }

        public static void Run(string filePath, Logger log)
        {
            var sol = new Solution1();
            var orbits = sol.GenerateInput(filePath);
            var orbitalSystem = sol.GenerateOrbitalMap(orbits);
            var total = sol.TotalOrbits(orbitalSystem);
            log.Information($"Total fuel required: {total}");
        }
    }
}